ugvc/pipelines/vcfeval_flavors.py

```diff
@@ -0,0 +1,113 @@
+                             'allowing transborder matches. To be used for truth-set', type=str)
+    ap.add_argument('-o', '--output',
+                        help='directory for output', type=str, required=True)
+    ap.add_argument('-t', '--template',
+                        help='SDF of the reference genome the variants are called against')
+    ap.add_argument('-p', '--allele_and_genotype_error_penalty',
+                        type=int,
+                        choices=[2, 1, 0, -1],
+                        default=0,
+                        help="-p 2: usual vcfeval, penalizes twice each wrong-allele/genotype errror\n" \
+                             "-p 1: penalize wrong-allele/genotype only once (remove 0.5 of such fps and fns)\n" \
+                             "-p 0: don't penalize wrong-allele/genotype (remove completely such fps and fns)\n" \
+                             "-p -1: don't penalize, and also reward a tp for calling wrong-allele/genotype\n")
+    return ap
+    return args
+
+def count_vcf_lines(vcf_file):
+    v_iter = pysam.VariantFile(vcf_file)
+    count = 0
+    for variant in v_iter:
+        count += 1
+    return count
+
+def run(argv: list[str]):
+    """Evaluate VCF against baseline, giving alternative penalty to wrong-alleles and genotype errors"""
+    parser = get_parser(argv)
+    SimplePipeline.add_parse_args(parser)
+    args = parser.parse_args(argv[1:])
+    sp = SimplePipeline(args.fc, args.lc, args.d)
+    calls = args.calls
+    baseline = args.baseline
+    eval_reg = args.evaluation_regions
+    sdf = args.template
+    out_dir = args.output
+    penalty = args.allele_and_genotype_error_penalty
+    result = []
+    sp.print_and_run(f'rm -rf {out_dir}')
+    sp.print_and_run(f'rtg vcfeval -b {baseline} -c {calls} -e {eval_reg} -t {sdf} -o {out_dir}')
+    sp.print_and_run(f'bedtools intersect -a {calls} -b {eval_reg} -header > {out_dir}/input_in_hcr.vcf')
+    sp.print_and_run(f'bcftools view -f PASS {out_dir}/input_in_hcr.vcf > {out_dir}/input_in_hcr.pass.vcf')
+    sp.print_and_run(f'bedtools intersect -a {baseline} -b {eval_reg} -header > {out_dir}/gtr_in_hcr.vcf')
+    result.append('type tp fp fn precision recall f1')
+    for vt in ['indels', 'snps']:
+        for pref in ['input_in_hcr', 'gtr_in_hcr', 'input_in_hcr.pass']:
+            sp.print_and_run(f'bcftools view {out_dir}/{pref}.vcf --type {vt} > {out_dir}/{pref}.{vt}.vcf')
+        sp.print_and_run(f'bcftools view {baseline} --type {vt} > {out_dir}/gt.{vt}.vcf')
+        for pref in ['fp', 'tp', 'fn']:
+            sp.print_and_run(f'bcftools view {out_dir}/{pref}.vcf.gz --type {vt} > {out_dir}/{pref}.{vt}.vcf')
+        sp.print_and_run(f'bedtools subtract -A -a {out_dir}/fp.{vt}.vcf -b {out_dir}/gt.{vt}.vcf -header >  {out_dir}/fp.{vt}.clean.vcf;')
+        sp.print_and_run(f'bedtools subtract -A -a {out_dir}/fn.{vt}.vcf -b {out_dir}/input_in_hcr.pass.{vt}.vcf -header > {out_dir}/fn.{vt}.clean.vcf')
```

Comment:
So you separate between "calling errors" and genotype errors, by looking at the same positions.
But what about multiallelic variants that appear in the same position? Or Cases where the ground truth variant has a different allele then what we have? (e.g. ground truth has AG insertion and we call AAG insertion, but they have the same position)

Also - consider using bcftools isec instead of bedtools

Reply:
I separate between calling errors (completely missing variant, or calling variant where there are none)
and other errors (genotype errors, or calling incorrect allele)